salt/migrations/2-3/haproxy/init.sls

```diff
@@ -0,0 +1,65 @@
+include:
+  - ca-cert
+  - cert
+
+# NOTE: Remove me for 4.0
+
+/etc/caasp/haproxy/haproxy.cfg:
+  file.managed:
+    - source: salt://haproxy/haproxy.cfg.jinja
+    - template: jinja
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+
+{% from '_macros/certs.jinja' import certs, alt_master_names with context %}
+{{ certs("kube-apiserver-proxy",
+         pillar['ssl']['kube_apiserver_proxy_crt'],
+         pillar['ssl']['kube_apiserver_proxy_key'],
+         cn = grains['nodename'] + '-proxy',
+         o = pillar['certificate_information']['subject_properties']['O'],
+         extra_alt_names = alt_master_names()) }}
+
+haproxy:
+  file.managed:
+    - name: /etc/kubernetes/manifests/haproxy.yaml
+    - source: salt://haproxy/haproxy.yaml.jinja
+    - template: jinja
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+  caasp_retriable.retry:
+    - name: iptables-haproxy
+{% if "kube-master" in salt['grains.get']('roles', []) %}
+    - target: iptables.append
+{% else %}
+    - target: iptables.delete
+{% endif %}
+    - retry:
+        attempts: 2
+    - table:      filter
+    - family:     ipv4
+    - chain:      INPUT
+    - jump:       ACCEPT
+    - match:      state
+    - connstate:  NEW
+    - dports:
+      - {{ pillar['api']['ssl_port'] }}
+    - proto:      tcp
+
+# Send a SIGTERM to haproxy when the config changes
+# TODO: There should be a better way to handle this, but currently, there is not. See
+# kubernetes/kubernetes#24957
+haproxy_restart:
+  cmd.run:
+    - name: |-
+        haproxy_id=$(docker ps | grep -E "k8s_haproxy.*\.{{ pillar['internal_infra_domain'] | replace(".", "\.") }}_kube-system_" | awk '{print $1}')
+        if [ -n "$haproxy_id" ]; then
+            docker kill $haproxy_id
```

Comment:
This is not going to work on clusters deployed with CRI-O, we should use `crictl stop` instead. Also the code that gets the container ID should be extended to handle CRI-O. We should have all the needed pieces inside of our `caasp_cri` salt module. If it's not there we should take a look at the file history, I'm pretty sure I wrote that...

Reply:
This is not going to run on clusters with cri-o. We need this state so it can be applied on 2.0 versions:

- Masters are updated
- Deploy new haproxy configuration to all workers (all still running 2.0)
- Old workers see the masters
- We can continue with the upgrade process on the workers (draining them etc...)

Reply:
I see, all good then!